### PR TITLE
Update MongoDB tests to not fail in Go 1.16

### PR DESCRIPTION
**Pull request by pcman312**
Go 1.16 introduced changes to the `x509.CertPool` type that make it incompatible with `reflect.DeepEqual` (https://github.com/golang/go/issues/45891). This caused some of our tests to start failing in MongoDB. This PR changes the comparison to use [github.com/google/go-cmp](https://github.com/google/go-cmp) for the equality check which allows us to ignore certain fields and types.

### CODE REVIEW

File: plugins/database/mongodb/mongodb_test.go [hunk 1]
```diff
@@ -363,6 +365,27 @@ func appendToCertPool(t *testing.T, pool *x509.CertPool, caPem []byte) *x509.Cer
...
+	cmp.AllowUnexported(x509.CertPool{}),
+}
+
+// Need a special comparison for ClientOptions because reflect.DeepEquals won't work in Go 1.16.
+// See: https://github.com/golang/go/issues/45891
+func assertDeepEqual(t *testing.T, a, b *options.ClientOptions) {
+	t.Helper()
+
+	if diff := cmp.Diff(a, b, cmpClientOptionsOpts); diff != "" {
+		t.Fatalf("assertion failed: values are not equal\n--- expected\n+++ actual\n%v", diff)
```
Comment (Valarissa) [inline plugins/database/mongodb/mongodb_test.go:385]: Could we see an output of this error message? Is it expecting no differences and then lists out the differences?

**Review by Valarissa — APPROVED** [review]
Looks good to me, but I'd love to see an example of the error message output!

File: plugins/database/mongodb/mongodb_test.go (hunk 1 above)
Comment (pcman312) [reply plugins/database/mongodb/mongodb_test.go:385]: Note: I flipped the assertion so it has the expected value on the left and the actual on the right.

Here's a failure if I change the `expectedOpts` in the `TestGetTLSAuth/good_ca` test to not include the provided `cert.Pem`:
Updated test case:
```
"good ca": {
	tlsCAData: cert.Pem,

	expectOpts: options.Client().
		SetTLSConfig(
			&tls.Config{
				RootCAs: x509.NewCertPool(), // <-- Note the missing CA
			},
		),
	expectErr: false,
},
```

Test output:
```
$ go test -v -run TestGetTLSAuth/good_ca
=== RUN   TestGetTLSAuth
=== RUN   TestGetTLSAuth/good_ca
    mongodb_test.go:353: assertion failed: values are not equal
        --- expected
        +++ actual
          &options.ClientOptions{
          	... // 22 identical fields
          	ServerSelectionTimeout: nil,
          	SocketTimeout:          nil,
          	TLSConfig: &tls.Config{
          		... // 7 identical fields
          		VerifyPeerCertificate: ⟪0x00⟫,
          		VerifyConnection:      ⟪0x00⟫,
        - 		RootCAs:               nil,
        + 		RootCAs: &x509.CertPool{
        + 			byName:    map[string][]int{"0\x141\x120\x10\x06\x03U\x04\x03\x13\ttest cert": {0}},
        + 			lazyCerts: []x509.lazyCert{{rawSubject: []uint8{...}, getCert: ⟪0x0124c580⟫}},
        + 			haveSum: map[x509.sum224]bool{
        + 				{0xbd, 0x0c, 0xeb, 0xa8, 0xfe, 0xaa, 0xe8, 0xaf, 0x6d, 0xdf, 0xb9, 0x82, 0x58, 0xff, 0x0c, 0x27, 0x3a, 0xbd, 0xd4, 0xba, 0x24, 0x8b, 0x47, 0xef, 0xbc, 0xbc, 0x78, 0xf0}: true,
        + 			},
        + 		},
          		NextProtos: nil,
          		ServerName: "",
          		... // 1 ignored and 16 identical fields
          	},
          	WriteConcern: nil,
          	ZlibLevel:    nil,
          	... // 6 identical fields
          }
--- FAIL: TestGetTLSAuth (0.44s)
    --- FAIL: TestGetTLSAuth/good_ca (0.00s)
FAIL
exit status 1
FAIL	github.com/hashicorp/vault/plugins/database/mongodb	0.881s
```

Same thing, but messing with the `good key` test:
```
"good key": {
			username:   "unittest",
			tlsKeyData: cert.CombinedPEM(),

			expectOpts: options.Client().
				SetTLSConfig(
					&tls.Config{}, // <-- Note the missing cert
				).
				SetAuth(options.Credential{
					AuthMechanism: "MONGODB-X509",
					Username:      "unittest",
				}),
			expectErr: false,
		},
```

**Comment by Valarissa** [comment]
Thanks for that update! Looks great!

**Review by calvn — APPROVED** [review]

**Review by austingebauer — APPROVED** [review]

**Pull request merged**